consensus/event_loop.go

```diff
@@ -0,0 +1,908 @@
+		fc.Events <- voteQCEvent
+	}
+
+	return nil
+}
+
+// ProcessValidatorTimeout captures an incoming timeout message from a validator. This module has no knowledge
+// of who the leader is for a given view, so it is up to the server to decide whether to process the timeout
+// message or not. If a timeout message is passed here, then the event loop will store it until
+// it can construct a QC with it or until the timeout's view has gone stale.
+//
+// This function does not directly check if the timeout results in a stake weighted super majority to build
+// a timeout QC. Instead, it stores the timeout locally and waits for the block production scheduled task to determine
+// when to run the super majority timeout check, and to signal the server that we can construct a timeout QC.
+//
+// Reference implementation:
+// https://github.com/deso-protocol/hotstuff_pseudocode/blob/6409b51c3a9a953b383e90619076887e9cebf38d/fast_hotstuff_bls.go#L958
+func (fc *fastHotStuffEventLoop) ProcessValidatorTimeout(timeout TimeoutMessage) error {
+	// Grab the event loop's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the event loop is running. This guarantees that the chain tip and validator list
+	// have already been set.
+	if fc.status != eventLoopStatusRunning {
+		return errors.New("FastHotStuffEventLoop.ProcessValidatorTimeout: Event loop is not running")
+	}
+
+	// Do a basic integrity check on the timeout message
+	if !isProperlyFormedTimeout(timeout) {
+		return errors.New("FastHotStuffEventLoop.ProcessValidatorTimeout: Malformed timeout message")
+	}
+
+	// Check if the timeout is stale
+	if isStaleView(fc.currentView, timeout.GetView()) {
+		return errors.Errorf("FastHotStuffEventLoop.ProcessValidatorTimeout: Timeout has a stale view %d", timeout.GetView())
+	}
+
+	// Check if the public key has already voted for this view. The protocol does not allow
+	// a validator to time out for a view it has already voted on.
+	if fc.hasVotedForView(timeout.GetPublicKey(), timeout.GetView()) {
+		return errors.Errorf(
+			"FastHotStuffEventLoop.ProcessValidatorTimeout: validator %s has already voted for view %d",
+			timeout.GetPublicKey().ToString(),
+			timeout.GetView(),
+		)
+	}
+
+	// Check if the public key has already timed out for this view. The protocol does not allow
+	// for a validator to time out more than once for the same view.
+	if fc.hasTimedOutForView(timeout.GetPublicKey(), timeout.GetView()) {
+		return errors.Errorf(
+			"FastHotStuffEventLoop.ProcessValidatorTimeout: validator %s has already timed out for view %d",
+			timeout.GetPublicKey().ToString(),
+			timeout.GetView(),
+		)
+	}
+
+	// Compute the value sha3-256(timeout.View, timeout.HighQC.View)
+	timeoutSignaturePayload := GetTimeoutSignaturePayload(timeout.GetView(), timeout.GetHighQC().GetView())
+
+	// Verify the vote signature
+	if !isValidSignatureSinglePublicKey(timeout.GetPublicKey(), timeout.GetSignature(), timeoutSignaturePayload[:]) {
```

Comment:
I could have just missed it, but I didn't see you validate the HighQC on the timeout message. I think you need to do this or else someone could attach a fake HighQC with a fake view number.